### PR TITLE
[WIP] Regressor trees

### LINKED ISSUE #28: [Easy, Python] Add Regressor tree and forest classes for Oblique and Patch trees
Add the regressor version of each type of tree.

### PR DESCRIPTION
**Pull request by SUKI-O**
Closes: #28 

Changes proposed in this pull request:

- Added ObliqueDecisionTreeRegressor
- Added PatchObliqueDecisionTreeRegressor
- Added ObliqueRadomForestRegressor
- Added PatchObliqueRadomForestRegressor

TODO:
- Add tests for ObliqueDecisionTreeRegressor
- Add tests for PatchObliqueDecisionTreeRegressor
- Add tests for ObliqueRadomForestRegressor
- Add test for PatchObliqueRadomForestRegressor

## Before submitting

- [ ] I've read and followed all steps in the [Making a pull request](https://github.com/py-why/pywhy-graphs/blob/main/CONTRIBUTING.md#making-a-pull-request)
    section of the `CONTRIBUTING` docs.
- [ ] I've updated or added any relevant docstrings following the syntax described in the
    [Writing docstrings](https://github.com/py-why/pywhy-graphs/blob/main/CONTRIBUTING.md#writing-docstrings) section of the `CONTRIBUTING` docs.
- [ ] If this PR fixes a bug, I've added a test that will fail without my fix.
- [ ] If this PR adds a new feature, I've added tests that sufficiently cover my new functionality.

## After submitting

- [ ] All GitHub Actions jobs for my pull request have passed.

### CODE REVIEW

**Comment by adam2392** [comment]
Hi @SUKI-O thanks for starting this! Can you resolve the conflicts? There have been some recent pushes to `main`, which made changes to the PatchObliqueTrees. This changes the API slightly

Lmk if you have any questions

File: sktree/tree/tests/test_tree.py [hunk 1]
```diff
@@ -125,6 +197,51 @@ def assert_tree_equal(d, s, message):
...
+    assert_allclose(reg.predict(T), y_test)
+
+    clf = Tree(criterion=criterion, max_features=1, random_state=1)
+    clf.fit(X, y_train)
+    assert_allclose(reg.predict(T), y_test)
+
+
 @parametrize_with_checks(
     [
         ObliqueDecisionTreeClassifier(random_state=12),
```
Comment (adam2392) [inline sktree/tree/tests/test_tree.py:182]: You'll want to add entries here. See: https://scikit-learn.org/stable/modules/generated/sklearn.utils.estimator_checks.check_estimator.html

Same w/ the forest versions in their respective test files.

Comment (SUKI-O) [reply sktree/tree/tests/test_tree.py:182]: Thank you so much for looking at this. I really appreciate it! 
Added the new regressor estimators. I'm sure I'm missing many other things here. Plz LMK if you see anything.

**Comment by adam2392** [comment]
The unit tests aren't passing, so you'll prolly need to debug those: https://github.com/neurodata/scikit-tree/actions/runs/4740945828/jobs/8417389584?pr=65

Same with code style formatting: https://github.com/neurodata/scikit-tree/actions/runs/4740945878/jobs/8417387360?pr=65

You'll also want to add a changelog entry: https://github.com/neurodata/scikit-tree/actions/runs/4740945860/jobs/8417387399?pr=65

You can run commands like (see pyproject.toml file):

```
poetry run poe format
poetry run poe lint
poetry run poe type_check
```

To run these you need `poetry` installed. Or you can run the checks manually. The poetry command is just an alias shortcut. 

Also to run unit-tests locally, you can use `spin`. I'll add this to the Contributing guide later

**Comment by SUKI-O** [comment]
> The unit tests aren't passing, so you'll prolly need to debug those: https://github.com/neurodata/scikit-tree/actions/runs/4740945828/jobs/8417389584?pr=65 
>
Question: In `test_check_importance_oblique`, you have 
```    
# The forest estimator can detect that only the first 3 features of the
    # dataset are informative:
    n_important = np.sum(importances > 0.1)
```
How is `0.1` determined? Is this something defined in sklearn? I couldn't find any reference. Or how is this test relevant in regressors? I tested with the original `RandomForestRegressor` from sklearn, but it also doesn't pass the test, either. For example, 
```
X = X_large_reg.astype(dtype, copy=False)
y = y_large_reg.astype(dtype, copy=False)

est = RandomForestRegressor(
    n_estimators=10,
    criterion="squared_error",
    random_state=0
)
est.fit(X, y)
importances = est.feature_importances_
print(importance)
------
[0.00948234 0.69881638 0.27146394 0.00296758 0.003216   0.00254183
 0.00314046 0.00305087 0.0030012  0.00231941]
```
I tested `ObliqueRandomForestRegressor` and `PatchObliqueRandomForestRegressor` and they all seem to have the first feature getting oddly low importance. Is this behavior expected? If so, have to come up with some other way to test this properly.

**Comment by SUKI-O** [comment]
I found to get around this issue: by setting `max_bins=3` when instantiating `ObliqueRandomForestRegressor`,  the importances generated become more heavily skewed and the test passes.  Do you see any problem with this?

File: sktree/tree/tests/test_tree.py [hunk 2]
```diff
@@ -439,3 +533,117 @@ def test_tree_deserialization_from_read_only_buffer(tmpdir, TREE):
...
+    # check consistency of trees when the depth and the number of features are
+    # limited
+
+    reg = Tree(criterion=criterion, max_depth=max_depth, max_features=6, random_state=0)
+    reg.fit(diabetes.data, diabetes.target)
+    loss = metric(diabetes.target, reg.predict(diabetes.data))
+    assert 0 < loss < max_loss
+
+
+def check_sparse_input(tree, dataset, max_depth=None):
```
Comment (adam2392) [inline sktree/tree/tests/test_tree.py]: I don't think this function is used, so perhaps delete? We currently don't have sparse support.

File: sktree/tests/test_supervised_forest.py [hunk 3]
```diff
@@ -469,20 +469,20 @@ def test_check_importances_oblique_regressor(criterion, dtype, feature_combinati
         n_estimators=10,
         criterion=criterion,
         random_state=0,
-        feature_combinations=feature_combinations,
+        max_bins=3,
```
Comment (adam2392) [inline sktree/tests/test_supervised_forest.py]: I would not touch max_bins as that is equivalent to making al the data into 3. I would also not touch the feature_combinations as that messes the existing test.

There is another test that you can add the regressor test into: `test_check_importances_oblique`.

File: sktree/tests/test_supervised_forest.py [hunk 4]
```diff
@@ -353,3 +391,185 @@ def test_check_importances_patch(criterion, dtype):
...
+
+    assert ri_mse < rc_mse
+    assert ri_mse > 0.85
+    assert rc_mse > 0.85
+
+
+def test_oblique_regressor_forest_trunk():
+    """Test oblique vs axis-aligned forests on Trunk."""
+    n = 1000
+    X, y = _trunk(n, p=100, random_state=0)
```
Comment (adam2392) [inline sktree/tests/test_supervised_forest.py]: Trunk and Orthant are classification problems, so these tests are not necessary

File: sktree/tests/test_supervised_forest.py [hunk 5]
```diff
@@ -353,3 +391,185 @@ def test_check_importances_patch(criterion, dtype):
...
+
+
+@pytest.mark.parametrize("dtype", (np.float64, np.float32))
+@pytest.mark.parametrize(
+    "criterion",
+    ["squared_error", "absolute_error", "friedman_mse"],
+)
+@pytest.mark.parametrize("feature_combinations", (2, 5))
+def test_check_importances_oblique_regressor(criterion, dtype, feature_combinations):
+    """Test checking feature importances for oblique trees."""
```
Comment (adam2392) [inline sktree/tests/test_supervised_forest.py]: You can add this test as a parametrization of the existing test ` test_check_importances_oblique`

File: sktree/tests/test_supervised_forest.py [hunk 6]
```diff
@@ -21,6 +34,29 @@
...
+}
+
+FOREST_REGRESSORS = {
+    "ObliqueRandomForestRegressor": ObliqueRandomForestRegressor,
+    "PatchObliqueRandomForestRegressor": PatchObliqueRandomForestRegressor,
+}
+
+FOREST_ESTIMATORS: Dict[str, Any] = dict()
+FOREST_ESTIMATORS.update(FOREST_CLASSIFIERS)
+FOREST_ESTIMATORS.update(FOREST_REGRESSORS)
```
Comment (adam2392) [inline sktree/tests/test_supervised_forest.py:60]: I don't see this used anywhere. Do we need this?

File: sktree/tests/test_supervised_forest.py [hunk 7]
```diff
@@ -21,6 +34,29 @@
...
+
+FOREST_CLASSIFIERS = {
+    "ObliqueRandomForestClassifier": ObliqueRandomForestClassifier,
+    "PatchObliqueRandomForestClassifier": PatchObliqueRandomForestClassifier,
+}
+
+FOREST_REGRESSORS = {
+    "ObliqueRandomForestRegressor": ObliqueRandomForestRegressor,
+    "PatchObliqueRandomForestRegressor": PatchObliqueRandomForestRegressor,
+}
```
Comment (adam2392) [inline sktree/tests/test_supervised_forest.py:56]: These aren't used anywhere it looks like?

File: sktree/tests/test_supervised_forest.py (hunk 6 above)
Comment (adam2392) [reply sktree/tests/test_supervised_forest.py:60]: Perhaps you could use them in the `test_sklearn_compatible_estimator` test?

**Review by adam2392 — CHANGES_REQUESTED** [review]
Okay it's looking better, but I would just take a look at the code you're adding to make sure if any code is added, it is actually used.

I think the rest of the PR is just some tidying up and then the main thing missing is a unit-test for regressors. Inside scikit-learn they use certain small simulated datasets to unit-test their RFRegressor and ExtreTreesRegressor. You can try using a similar set of datasets to unit-test the ObliqueForestRegressor.

The Patch ObliqueRegressor tho is a bit more tricky because it is meant for image like datasets, so your `X` needs to be images, but your `y` needs to be some continuous variable. Idk if this is well defined, so you can perhaps not add the patchobliqueRFRegressor class. Or, feel free to think of a test case that makes sense.

File: CONTRIBUTING.md
Comment (adam2392) [inline CONTRIBUTING.md:1]: Where did this change come from?

File: README.md
Comment (adam2392) [inline README.md:1]: Same here?

File: CONTRIBUTING.md
Comment (adam2392) [reply CONTRIBUTING.md:1]: Maybe you rebased off of a PR branch I made somewhere else? Try to always base a PR off of `main`, so it's easier to track what PR changes what.

File: sktree/tests/test_supervised_forest.py (hunk 4 above)
Comment (SUKI-O) [reply sktree/tests/test_supervised_forest.py]: Copy

File: sktree/tests/test_supervised_forest.py (hunk 7 above)
Comment (SUKI-O) [reply sktree/tests/test_supervised_forest.py:56]: I thought about using these to factor some of the repeated tests for classifier & regressor then I realized I defined wrong but I fixed it. Thought about using it in `test_check_importances_oblique`, `test_check_importances_patch` but not sure if works.  But yes, can be used in `test_sklearn_compatible_estimator`

**Comment by SUKI-O** [comment]
> 

> Okay it's looking better, but I would just take a look at the code you're adding to make sure if any code is added, it is actually used.
> 
> I think the rest of the PR is just some tidying up and then the main thing missing is a unit-test for regressors. Inside scikit-learn they use certain small simulated datasets to unit-test their RFRegressor and ExtreTreesRegressor. You can try using a similar set of datasets to unit-test the ObliqueForestRegressor.
> 
> The Patch ObliqueRegressor tho is a bit more tricky because it is meant for image like datasets, so your `X` needs to be images, but your `y` needs to be some continuous variable. Idk if this is well defined, so you can perhaps not add the patchobliqueRFRegressor class. Or, feel free to think of a test case that makes sense.

Do you mean I need to add unit tests for `ObliqueRandomForestRegressor` and `PatchObliqueRandomForestRegressor` with a toy set? I didn't realize that `Patch` one is intended for images but makes sense. But is it exclusive? Do the inputs have to be 2D images? They both pass the unit test I just tried. Do you think we should not include `PatchObliqueRandomForestRegressor` anyways?

**Comment by SUKI-O** [comment]
Meson builds are failing. Getting the errors that look like:
`/home/runner/work/scikit-tree/scikit-tree/sktree/tree/_unsup_criterion.pxd:6:0: 'sklearn/tree/_criterion.pxd' not found`
Did I mess up something in Meson?

**Comment by adam2392** [comment]
Nah Dw about those builds. I'm changing some stuff in the fork. But I'll fix it soon

**Comment by adam2392** [comment]
> 
> Do you mean I need to add unit tests for `ObliqueRandomForestRegressor` and `PatchObliqueRandomForestRegressor` with a toy set? I didn't realize that `Patch` one is intended for images but makes sense. But is it exclusive? Do the inputs have to be 2D images? They both pass the unit test I just tried. Do you think we should not include `PatchObliqueRandomForestRegressor` anyways?

Good question! It can be any dataset with correlations within a patch. Images, time series, maybe correlated "nearby" features. Doesn't have to be images. The best would be some dataset that has structure with y being continuous. Idk if there is one from sklearn easily accessible. 

Yes adding a toy dataset with a unit test to assert some type of performance serves as a regression test against future changes.

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -1,7 +1,7 @@
 # Contributing
 
 Thanks for considering contributing! Please read this document to learn the various ways you can contribute to this project and how to go about doing it.
-
+`
```
Comment (adam2392) [inline CONTRIBUTING.md:4]:
```suggestion

```

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -1,7 +1,6 @@
 # Contributing
 
 Thanks for considering contributing! Please read this document to learn the various ways you can contribute to this project and how to go about doing it.
-
 ## Bug reports and feature requests
```
Comment (adam2392) [inline CONTRIBUTING.md:5]:
```suggestion

## Bug reports and feature requests
```

File: examples/plot_projection_matrices.py
Comment (adam2392) [inline examples/plot_projection_matrices.py:1]: I think you rebased off of #67 instead of `main` perhaps.

File: sktree/tree/_classes.py [hunk 10]
```diff
@@ -1164,7 +2084,7 @@ def __init__(
         random_state=None,
         max_leaf_nodes=None,
         min_impurity_decrease=0.0,
-        class_weight=None,
+        ccp_alpha=0.0,
```
Comment (adam2392) [inline sktree/tree/_classes.py]: We don't implement cost-complexity pruning

File: examples/plot_projection_matrices.py
Comment (adam2392) [reply examples/plot_projection_matrices.py:1]: I think tho this is the only "extra change", so feel free to just remove this change so it doesn't show up in the PR.

**Review by adam2392 — CHANGES_REQUESTED** [review]
Thanks for adding unit tests! If you can think of a good unit-test for the regression case of PatchObliqueForests, feel free to add and open up a discussion thread. It'll be important forsure.

The CIs might fail for now, so I would check all the formatting, linting and docs build locally. Once the CIs are fixed I'll re-run things here. That + if you fix up some of the comments I left, then this should be good to merge.

Once you've addressed everything, can you "Resolve conversation" for all the comments on the PR and then mark this PR "Ready for Review"?

File: docs/whats_new/v0.1.rst [hunk 11]
```diff
@@ -31,6 +31,7 @@ Changelog
 - |Feature| Implementation of fastBIC criterion for unsupervised tree models, by `Adam Li`_ and `Jong Shin`_ (:pr:`45`)
 - |Fix| Fix a bug in Patch oblique random forest that samples outside the data boundaries and adds a user guide, by `Adam Li`_ (:pr:`61`)
 - |Feature| MORF trees now can sample n-dimensional patches inside an n-dimensional structure sample and make any arbitrary axis discontinuous, by `Adam Li`_ (:pr:`63`)
+- |Feature| Implementation of ObliqueDecisionTreeRegressor and PatchObliqueDecisionTreeRegressor, by `SUKI-O`_ (:pr:`65`)
```
Comment (adam2392) [inline docs/whats_new/v0.1.rst:35]: Add your name to the `authors` file in the same format as I added mine

File: examples/plot_projection_matrices.py
Comment (SUKI-O) [reply examples/plot_projection_matrices.py:1]: I swear I rebased it to `main`.. but, sorry for not being careful.

Comment (adam2392) [reply examples/plot_projection_matrices.py:1]: No worries! Happens all the time :)

**Comment by adam2392** [comment]
> Nah Dw about those builds. I'm changing some stuff in the fork. But I'll fix it soon

This should now be fixed on `main`. See #67 for details.

**Comment by adam2392** [comment]
I think instead of rebasing in the future, it'll be easier to merge commit changes from main. For now I think there's a bunch of stuff that got messed up on rebase. 

Another option is copying relevant changes into a new PR branch based off of main.

**Pull request closed without merging**